File: src/ast.c
/*
 * XENLY - high-level and general-purpose programming language
 * created, designed, and developed by Cyril John Magayaga (cjmagayaga957@gmail.com, cyrilmagayaga@proton.me).
 *
 * It is initially written in C programming language.
 *
 * It is available for Linux and macOS operating systems.
 *
 */
#include "ast.h"
#include <stdlib.h>
#include <string.h>
#include <stdio.h>

// ─── Create ──────────────────────────────────────────────────────────────────
ASTNode *ast_node_create(NodeType type, int line) {
    ASTNode *n       = (ASTNode *)calloc(1, sizeof(ASTNode));
    n->type          = type;
    n->line          = line;
    n->child_capacity = 4;
    n->children      = (ASTNode **)malloc(sizeof(ASTNode *) * n->child_capacity);
    return n;
}

// ─── Add Child ───────────────────────────────────────────────────────────────
void ast_node_add_child(ASTNode *parent, ASTNode *child) {
    if (parent->child_count >= parent->child_capacity) {
        parent->child_capacity *= 2;
        parent->children = (ASTNode **)realloc(
            parent->children,
            sizeof(ASTNode *) * parent->child_capacity
        );
    }
    parent->children[parent->child_count++] = child;
}

// ─── Destroy (recursive) ─────────────────────────────────────────────────────
void ast_node_destroy(ASTNode *node) {
    if (!node) return;
    // Recursively destroy children
    for (size_t i = 0; i < node->child_count; i++)
        ast_node_destroy(node->children[i]);
    free(node->children);
    // Free params
    if (node->params) {
        for (size_t i = 0; i < node->param_count; i++)
            free(node->params[i].name);
        free(node->params);
    }
    free(node->str_value);
    free(node);
}

// ─── Debug Print ─────────────────────────────────────────────────────────────
static const char *node_type_name(NodeType t) {
    switch(t) {
        case NODE_PROGRAM:        return "PROGRAM";
        case NODE_VAR_DECL:       return "VAR_DECL";
        case NODE_ASSIGN:         return "ASSIGN";
        case NODE_COMPOUND_ASSIGN:return "COMPOUND_ASSIGN";
        case NODE_INCREMENT:      return "INCREMENT";
        case NODE_DECREMENT:      return "DECREMENT";
        case NODE_FN_DECL:        return "FN_DECL";
        case NODE_FN_CALL:        return "FN_CALL";
        case NODE_RETURN:         return "RETURN";
        case NODE_IF:             return "IF";
        case NODE_WHILE:          return "WHILE";
        case NODE_PRINT:          return "PRINT";
        case NODE_INPUT:          return "INPUT";
        case NODE_IMPORT:         return "IMPORT";
        case NODE_BLOCK:          return "BLOCK";
        case NODE_BINARY:         return "BINARY";
        case NODE_UNARY:          return "UNARY";
        case NODE_NUMBER:         return "NUMBER";
        case NODE_STRING:         return "STRING";
        case NODE_BOOL:           return "BOOL";
        case NODE_NULL:           return "NULL";
        case NODE_IDENTIFIER:     return "IDENT";
        case NODE_METHOD_CALL:    return "METHOD_CALL";
        case NODE_EXPR_STMT:      return "EXPR_STMT";
        case NODE_CLASS_DECL:     return "CLASS_DECL";
        case NODE_NEW:            return "NEW";
        case NODE_THIS:           return "THIS";
        case NODE_SUPER_CALL:     return "SUPER_CALL";
        case NODE_PROPERTY_GET:   return "PROP_GET";
        case NODE_PROPERTY_SET:   return "PROP_SET";
        case NODE_TYPEOF:         return "TYPEOF";
        case NODE_INSTANCEOF:     return "INSTANCEOF";
        case NODE_CALL_EXPR:      return "CALL_EXPR";
        default:                  return "???";
    }
}

void ast_print(ASTNode *node, int indent) {
    if (!node) return;
    for (int i = 0; i < indent; i++) printf("  ");

    printf("[%s]", node_type_name(node->type));

    if (node->str_value)  printf(" str=\"%s\"", node->str_value);
    if (node->type == NODE_NUMBER) printf(" num=%.6g", node->num_value);
    if (node->type == NODE_BOOL)   printf(" bool=%s", node->bool_value ? "true" : "false");
    if (node->type == NODE_FN_DECL && node->params) {
        printf(" params=(");
        for (size_t i = 0; i < node->param_count; i++) {
            if (i) printf(", ");
            printf("%s", node->params[i].name);
        }
        printf(")");
    }
    printf("\n");

    for (size_t i = 0; i < node->child_count; i++)
        ast_print(node->children[i], indent + 1);
}

File: xvm/pkg/compiler/ast.go
/*
 * XENLY VIRTUAL MACHINE (XVM) - high-performance of the Virtual Machine
 * created, designed, and developed by Cyril John Magayaga (cjmagayaga957@gmail.com, cyrilmagayaga@proton.me).
 *
 * It is initially written in Go programming language.
 *
 * It is available for the Linux, macOS, and Windows operating systems.
 *
 */
/*
 * XENLY - Xenly Virtual Machine (XVM)
 * Compiler: AST node definitions
 */
package compiler

// NodeKind identifies the kind of AST node.
type NodeKind int

const (
	// ── Program / Declarations ────────────────────────────────────────────
	NodeProgram     NodeKind = iota // root
	NodeVarDecl                     // var/let x = expr
	NodeConstDecl                   // const x = expr
	NodeFnDecl                      // fn name(params) { body }
	NodeClassDecl                   // class Name [extends P] { ... }
	NodeEnumDecl                    // enum Name { Variant(fields), ... }
	NodeTypeAlias                   // type Alias = Type
	NodeNamespace                   // namespace Name { ... }
	NodeExport                      // export fn/class
	NodeImport                      // import "mod" [as alias] | from "mod" import ...

	// ── Statements ────────────────────────────────────────────────────────
	NodeBlock        // { stmts }
	NodeExprStmt     // expression used as statement
	NodeAssign       // x = expr
	NodeCompoundAssign // x += expr etc.
	NodeReturn       // return [expr]
	NodeIf           // if (cond) body [else body]
	NodeWhile        // while (cond) body
	NodeDoWhile      // do body while (cond)
	NodeFor          // for (init; cond; update) body
	NodeForIn        // for (var x in expr) body
	NodeSwitch       // switch (expr) { case val: stmts... }
	NodeBreak        // break
	NodeContinue     // continue
	NodeIncrement    // x++
	NodeDecrement    // x--
	NodePrint        // print(args...)
	NodeInput        // input([prompt])
	NodeSleep        // sleep(ms)
	NodeAssert       // assert(cond, msg)
	NodeRequires     // requires (cond)
	NodeEnsures      // ensures (cond)
	NodeInvariant    // invariant (cond)
	NodeSpawn        // spawn expr
	NodeAwait        // await expr

	// ── Expressions ───────────────────────────────────────────────────────
	NodeBinary        // left op right
	NodeUnary         // op right
	NodeTernary       // cond ? t : f
	NodeNullish       // a ?? b
	NodeNumber        // 42
	NodeString        // "hello"
	NodeBool          // true/false
	NodeNull          // null
	NodeIdent         // varName
	NodeFnCall        // fn(args)
	NodeMethodCall    // obj.method(args)
	NodeCallExpr      // expr(args)  — call any expression
	NodePropertyGet   // obj.prop
	NodePropertySet   // obj.prop = val
	NodeIndex         // arr[idx]
	NodeIndexAssign   // arr[idx] = val
	NodeThis          // this
	NodeSuper         // super(args)
	NodeNew           // new ClassName(args)
	NodeTypeof        // typeof expr
	NodeInstanceof    // expr instanceof Class
	NodeArrayLiteral  // [items...]
	NodeObjectLiteral // {key: val, ...}
	NodeArrowFn       // (params) => expr
	NodeEnumVariant   // Tag(fields)
	NodeMatch         // match expr { arms... }
	NodeMatchArm      // Pattern => expr
	NodePattern       // match pattern
	NodeNamedArg      // name: value (named argument)
	NodeTupleLiteral  // (a, b, c)
	NodeNamespaceAccess // Ns.member
	NodeSpread        // ...expr
)

// ASTNode is a node in the abstract syntax tree.
type ASTNode struct {
	Kind     NodeKind
	Line     int
	// Shared value fields
	StrVal   string  // identifier, string literal, operator symbol
	NumVal   float64 // numeric literal
	BoolVal  bool    // bool literal

	// Children
	Children []*ASTNode

	// Function / arrow function parameters
	Params []*FnParam

	// Type info (for gradual typing)
	TypeAnnotation string
	ReturnType     string

	// Extra flags
	IsAsync  bool // async fn
	IsExport bool // export declaration
}

// FnParam represents a function parameter.
type FnParam struct {
	Name         string
	TypeAnnot    string
	DefaultValue *ASTNode // may be nil
	IsRest       bool     // ...rest param
	IsOptional   bool
}

// Child helpers

func (n *ASTNode) AddChild(child *ASTNode) {
	n.Children = append(n.Children, child)
}

func NewNode(kind NodeKind, line int) *ASTNode {
	return &ASTNode{Kind: kind, Line: line}
}

func NumberNode(val float64, line int) *ASTNode {
	n := NewNode(NodeNumber, line)
	n.NumVal = val
	return n
}

func StringNode(val string, line int) *ASTNode {
	n := NewNode(NodeString, line)
	n.StrVal = val
	return n
}

func BoolNode(val bool, line int) *ASTNode {
	n := NewNode(NodeBool, line)
	n.BoolVal = val
	return n
}

func IdentNode(name string, line int) *ASTNode {
	n := NewNode(NodeIdent, line)
	n.StrVal = name
	return n
}
